helper/schema/grpc_provider.go

```diff
@@ -607,12 +607,37 @@ func (s *GRPCProviderServer) ConfigureProvider(ctx context.Context, req *tfproto
 	// request scoped contexts, however this is a large undertaking for very large providers.
 	ctxHack := context.WithValue(ctx, StopContextKey, s.StopContext(context.Background()))
 
+	// NOTE: This is a hack to pass the deferral_allowed field from the Terraform client to the
+	// underlying (provider).Configure function, which cannot be changed because the function
+	// signature is public. (╯°□°)╯︵ ┻━┻
+	s.provider.deferralAllowed = configureDeferralAllowed(req.ClientCapabilities)
+
 	logging.HelperSchemaTrace(ctx, "Calling downstream")
 	diags := s.provider.Configure(ctxHack, config)
 	logging.HelperSchemaTrace(ctx, "Called downstream")
 
 	resp.Diagnostics = convert.AppendProtoDiag(ctx, resp.Diagnostics, diags)
 
+	if s.provider.providerDeferred != nil {
+		// Check if a deferred response was incorrectly set on the provider. This would cause an error during later RPCs.
+		if !configureDeferralAllowed(req.ClientCapabilities) {
+			resp.Diagnostics = append(resp.Diagnostics, &tfprotov5.Diagnostic{
+				Severity: tfprotov5.DiagnosticSeverityError,
+				Summary:  "Invalid Deferred Provider Response",
+				Detail: "Provider configured a deferred response for all resources and data sources but the Terraform request " +
+					"did not indicate support for deferred actions. This is an issue with the provider and should be reported to the provider developers.",
+			})
+		} else {
+			logging.HelperSchemaDebug(
+				ctx,
+				"Provider has configured a deferred response, all associated resources and data sources will automatically return a deferred response.",
+				map[string]interface{}{
+					"deferred_reason": s.provider.providerDeferred.Reason.String(),
```

Comment:
Nit: Should this logging key be `tf_deferred_reason` to match terraform-plugin-go here? https://github.com/hashicorp/terraform-plugin-go/blob/05a54cd0f10761f941562bbf8846932d551e0c99/internal/logging/keys.go#L67

This Go module does have an internal constants file for logging keys, if that helps! https://github.com/hashicorp/terraform-plugin-sdk/blob/main/internal/logging/keys.go

Same with the other usage of this key.

Reply:
Thanks for catching that! Added it to the constants -> bd21c5e19779c42be5748e84b62dff2bd77cee01 ⚡